pkg/proc/native/proc.go

```diff
@@ -333,20 +335,48 @@ func (dbp *nativeProcess) handlePtraceFuncs() {
 	runtime.LockOSThread()
 
 	for fn := range dbp.ptraceChan {
+		dbp.insidePtraceThread = true
 		fn()
+		dbp.insidePtraceThread = false
 		dbp.ptraceDoneChan <- nil
 	}
+	close(dbp.ptraceDoneChan)
 }
 
 func (dbp *nativeProcess) execPtraceFunc(fn func()) {
+	if dbp.insidePtraceThread {
+		fn()
+		return
+	}
 	dbp.ptraceChan <- fn
 	<-dbp.ptraceDoneChan
 }
 
+// ExecOnMagicThread executes fn on the thread that can make ptrace
+// calls (or the equivalent of ptrace on windows). If fn does a lot of calls
+// to other methods of Process that need to do ptrace calls this will speed
+// it up considerably by avoiding a large number of context switches.
+// It is very important that fn does not panic, you must defer a recover
+// call here.
+//
+// This function works by calling fn in the ptrace goroutine and then
+// disabling the code that sends ptrace calls to that goroutine, executing
+// them directly instead.
+// It is very important that only the goroutine that calls this function
```

Comment:
Since we no longer have the guarantee that the calls will be serialized we need to enforce it and ensure we produce correct behavior in some other way than relying on the channel to serialize.

I'd be curious to see the performance implications of ensuring we're executing on the correct thread. E.g. using `syscall.Gettid()` and checking against the thread we know to be the "ptrace" or "magic" thread.